src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs

```diff
@@ -1,93 +1,604 @@
+
+    public override string ToString() => _handle.ToString();
+
+    /// <summary>
+    /// Initializes a new instance of an outbound <see cref="QuicConnection" />.
+    /// </summary>
+    private unsafe QuicConnection()
+    {
+        GCHandle context = GCHandle.Alloc(this, GCHandleType.Weak);
+        try
+        {
+            QUIC_HANDLE* handle;
+            ThrowIfFailure(MsQuicApi.Api.ApiTable->ConnectionOpen(
+                MsQuicApi.Api.Registration.QuicHandle,
+                &NativeCallback,
+                (void*)GCHandle.ToIntPtr(context),
+                &handle));
+            _handle = new MsQuicContextSafeHandle(handle, context, MsQuicApi.Api.ApiTable->ConnectionClose, SafeHandleType.Connection);
+        }
+        catch
+        {
+            context.Free();
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// Initializes a new instance of an inbound <see cref="QuicConnection" />.
+    /// </summary>
+    /// <param name="handle">Native handle.</param>
+    /// <param name="info">Related data from the NEW_CONNECTION listener event.</param>
+    internal unsafe QuicConnection(QUIC_HANDLE* handle, QUIC_NEW_CONNECTION_INFO* info)
+    {
+        GCHandle context = GCHandle.Alloc(this, GCHandleType.Weak);
+        try
+        {
+            delegate* unmanaged[Cdecl]<QUIC_HANDLE*, void*, QUIC_CONNECTION_EVENT*, int> nativeCallback = &NativeCallback;
+            MsQuicApi.Api.ApiTable->SetCallbackHandler(
+                handle,
+                nativeCallback,
+                (void*)GCHandle.ToIntPtr(context));
+            _handle = new MsQuicContextSafeHandle(handle, context, MsQuicApi.Api.ApiTable->ConnectionClose, SafeHandleType.Connection);
+        }
+        catch
+        {
+            context.Free();
+            throw;
+        }
+
+        _remoteEndPoint = info->RemoteAddress->ToIPEndPoint();
+        _localEndPoint = info->LocalAddress->ToIPEndPoint();
+    }
+
+    private async ValueTask FinishConnectAsync(QuicClientConnectionOptions options, CancellationToken cancellationToken = default)
     {
-        public static bool IsSupported => MsQuicApi.IsQuicSupported;
+        ObjectDisposedException.ThrowIf(_disposed == 1, this);
 
-        public static ValueTask<QuicConnection> ConnectAsync(QuicClientConnectionOptions options, CancellationToken cancellationToken = default)
+        if (_connectedTcs.TryInitialize(out ValueTask valueTask, this, cancellationToken))
         {
-            if (!IsSupported)
+            _canAccept = options.MaxInboundBidirectionalStreams > 0 || options.MaxInboundUnidirectionalStreams > 0;
+            _defaultCloseErrorCode = options.DefaultCloseErrorCode;
+
+            if (!options.RemoteEndPoint.TryParse(out string? host, out IPAddress? address, out int port))
+            {
+                throw new ArgumentException($"Unsupported remote endpoint type '{options.RemoteEndPoint.GetType()}', expected IP or DNS endpoint.", nameof(options));
```

Comment:
This should go to localizable string resources.

Reply:
Fixed